htdocs/product/class/html.formproduct.class.php

```diff
@@ -329,5 +330,126 @@ function load_measuring_units($name='measuring_units', $measuring_style='', $def
+
+		$out='';
+
+		$nboflot = $this->loadLotStock($fk_product, $fk_entrepot);
+
+		if ($conf->use_javascript_ajax && ! $forcecombo)
+		{
+			include_once DOL_DOCUMENT_ROOT . '/core/lib/ajax.lib.php';
+			$comboenhancement = ajax_combobox($htmlname, $events);
+			$out.= $comboenhancement;
+		}
 
+		$out.='<select class="flat'.($morecss?' '.$morecss:'').'"'.($disabled?' disabled':'').' id="'.$htmlname.'" name="'.($htmlname.($disabled?'_disabled':'')).'">';
+		if ($empty) $out.='<option value="-1">'.($empty_label?$empty_label:'&nbsp;').'</option>';
+		foreach($this->cache_lot as $id => $arraytypes)
+		{
+			$out.='<option value="'.$id.'"';
+			if ($selected == $id || ($selected == 'ifone' && $nboflot == 1)) $out.=' selected';
+			$out.='>';
+			$out.=$arraytypes['entrepot_label'].' - ';
+			$out.=$arraytypes['batch'];
+			$out.=' ('.$langs->trans("Stock").':'.$arraytypes['qty'].')';
+			$out.='</option>';
+		}
+		$out.='</select>';
+		if ($disabled) $out.='<input type="hidden" name="'.$htmlname.'" value="'.(($selected>0)?$selected:'').'">';
+
+		return $out;
+	}
+
+	/**
+	 * Load in cache array list of lot available in stock
+	 * If fk_product is not 0, we do not use cache
+	 *
+	 * @param	int		$fk_product		    load lot of id fk_product, all if 0. 
+	 * @param	string	$fk_entrepot	    load lot in fk_entrepot all if 0.
+	 *
+	 * @return  int  		    		    Nb of loaded lines, 0 if already loaded, <0 if KO
+	 */
+	function loadLotStock($fk_product = 0, $fk_entrepot = 0)
+	{
+		global $conf, $langs;
+
+		if (empty($fk_product) && empty($fk_product) && count($this->cache_lot))
+		{
+			return count($this->cache_lot); // Cache already loaded and we do not want a list with information specific to a product or warehouse
+		}
+		else 
+		{
+			// clear cache;
+			$this->cache_lot = array();
+		}
+
+		$sql = "SELECT pb.batch, pb.rowid, ps.fk_entrepot, pb.qty, e.label";
+		$sql.= " FROM ".MAIN_DB_PREFIX."product_batch as pb";
+		$sql.= " LEFT JOIN ".MAIN_DB_PREFIX."product_stock as ps on ps.rowid = pb.fk_product_stock";
+		$sql.= " LEFT JOIN ".MAIN_DB_PREFIX."entrepot as e on e.rowid = ps.fk_entrepot AND e.entity IN (".getEntity('stock').")";
+		if (!empty($fk_product) || !empty($fk_entrepot))
+		{
+			$sql.= " WHERE";
+			if (!empty($fk_product))
+			{
+				$sql.= " ps.fk_product = '".$fk_product."'";
+				if (!empty($fk_entrepot))
+				{
+					$sql.= " AND";
+				}
+			}
+			if (!empty($fk_entrepot))
+			{
+				$sql.= " ps.fk_entrepot = '".$fk_entrepot."'";
+			}
+		}
+		$sql.= " ORDER BY e.label, pb.batch";
+
+		dol_syslog(get_class($this).'::loadLotStock', LOG_DEBUG);
+		$resql = $this->db->query($sql);
+		if ($resql)
+		{
+			$num = $this->db->num_rows($resql);
+			$i = 0;
+			while ($i < $num)
+			{
+				$obj = $this->db->fetch_object($resql);
+				$this->cache_lot[$obj->rowid]['id'] =$obj->rowid;
```

Comment:
If fk_product or fk_entrepot are set at first call, we load data restricted to product or warehouse. And the cache array contains only this restricted data.
If we wall again the method loadLotStock with empty fk_product and empty fk_entrepot, then it return the cache. But this cache does not contains all data, so content is not complete for the second call. I suggest to add a key into cache
$this->cache_lot[fk_product.'_'.fk_warehouse]

Reply:
I will check this, but will need to consider memory/performance. A company can have a lot off different lot/serial numbers in stock.

Reply:
I have added a new parameter to be able to cash all lot stock for all products in lines. So you could also use this new form select for creating lines. I don't see any use case of loading all lot numbers for all products.